tests/main/security-device-cgroups-strict-enforced/task.yaml

```diff
@@ -47,3 +51,30 @@ execute: |
     not snap run test-strict-cgroup.sh -c 'dd if=/dev/zero of=/dev/full bs=1 count=1' > run.log 2>&1
     # and dd correctly fails observing ENOSPC
     MATCH "dd: error writing '/dev/full': No space left on device" < run.log
+
+    if os.query is-trusty; then
+        # next part requires being able to trigger udev event related to
+        # specific device but udevadm 14.04 has no switches for that
+        exit 0
+    fi
+
+    rm -f /var/snap/test-strict-cgroup/common/started
+    rm -f /var/snap/test-strict-cgroup/common/ready
+
+    # trigger a change event for /dev/full so that we observe that the device
+    # cgroup settings remain stable and we can still access /dev/zero
+    snap run test-strict-cgroup.sh -c 'touch /var/snap/test-strict-cgroup/common/started; until test -e /var/snap/test-strict-cgroup/common/ready; do sleep 1; done; echo ok > /dev/zero' > run.log 2>&1 &
```

Comment:
I'm not sure that this spread test is testing the same thing that actually failed on the device in question, IIUC, what was happening is something like this order:

1. new udev event triggers snap-device-helper to start running
2. concurrently, some script is executing trying to access /dev/null (but hasn't accessed it yet)
3. snap-device-helper re-sets the device cgroup to deny everything
4. the aforementioned script tries to access /dev/null and fails since the cgroup denies everything
5. snap-device-helper finishes setting up the device cgroup and now /dev/null is allowed again
6. subsequent restarts of the script don't fail and /dev/null is in the device cgroup and allowed, etc

whereas this test seems to be doing this order:

1. some script starts executing trying to access /dev/null (but hasn't accessed it yet)
2. udev event is triggered to make snap-device-helper start running
3. snap-device-helper finishes setting up the device cgroup
4. the script is notified to try and access /dev/null again, and it does so
5. no errors

The chief difference to me is that in the spread test we are attempting to check the race condition at all, since we are accessing the device only after snap-device-helper is done executing, which AIUI that test case would pass even without this PR on master as well, since the device cgroup changes after snap-device-helper is done apply to running processes same as new processes, etc.

Reply:
I would image a spread test which does attempt to emulate the race condition (to the best of it's abilities) is a script like:

```
    snap run test-strict-cgroup.sh -c 'while true; do echo ok > /dev/zero; done' > run.log 2>&1 &
    pid=$!
    # do the udev trigger, saving the cgroup state before/after
    # check that there was no denial and that the program is still running/active and didn't exit non-zero
    # diff the cgroup before/after make sure it's the same
    # kill the process we forked above
```

Reply:
Not quite,
1. snap app starts executing (device cgroup has been created at this point allowing access to specific device + the common set)
2. udev event triggers snap-device-helper
3. snap-device-helper only amends the device cgroup for that device
4. the app continues execution accessing a device that is in the common set

The test did not pass on master as on v1 all devices were wiped due to the bug and only `c 1:7` was added to the allowed set, thus accessing /dev/zero (c 1:3) failed.

Reply:
Fwiw, I did ran this test against the unmodified snap-confine and got:
```
+ retry -n 5 test -e /var/snap/test-strict-cgroup/common/started
+ snap run test-strict-cgroup.sh -c 'touch /var/snap/test-strict-cgroup/common/started; until test -e /var/snap/test-strict-cgroup/common/ready; do sleep 1; done; echo ok > /dev/zero'
retry: command test -e /var/snap/test-strict-cgroup/common/started failed with code 1
retry: next attempt in 1 second(s) (attempt 1 of 5)
+ sort
+ tests.device-cgroup test-strict-cgroup.sh dump
+ udevadm trigger --name-match /dev/full
+ udevadm settle
+ sort
+ tests.device-cgroup test-strict-cgroup.sh dump
+ touch /var/snap/test-strict-cgroup/common/ready
+ wait
+ MATCH '^$'
grep error: pattern not found, got:
/bin/sh: 1: cannot create /dev/zero: Operation not permitted
```
but maybe I'm missing something?

Reply:
Maybe worth adding some extra comment here though (and if-so probably best in a skip-spread followup :)

Reply:
So really in v1 with master right now (i.e. not with this patch), if a program was running accessing devices, and a new udev event was triggered that added a device to the device cgroup for that program, that program would _only_ have access to the device that was added after snap-device-helper ran? If so, indeed that is arguably much worse than I thought the bug was. 

I didn't realize that was the case because it wasn't quite what Alfonso explained in practice on the device, where after the event was processed the device cgroup for the snap contained all the devices. But perhaps what was happening was that the code would clear the device cgroup for the existing process and add that one device to just the device cgroup the existing process was in, but then all other new processes which are started afterwards would create a fresh new cgroup with all the devices (the common set + the new device) so then the commands that Alfonso was running would be executing in the full, expected cgroup and thus not be denied. Is that what was happening then?

Reply:
I'll add a comment in a followup